app/services/foreman/template_snapshot_service.rb

```diff
@@ -34,33 +72,132 @@ def host
-        FactoryBot.build(:host_parameter, host: host, name: name, value: value)
-      end
-      host.define_singleton_method(:params) { host_params }
-      host.define_singleton_method(:host_param) do |name|
-        host_params[name]
-      end
-      host
+    def host4static
+      interface = FactoryBot.build(:nic_primary_and_provision, identifier: 'eth0',
+        mac: '00-f0-54-1a-7e-e0',
+        ip: '192.168.42.42')
+      domain = FactoryBot.build(:domain, name: 'snapshotdomain.com')
+      subnet = FactoryBot.build(:subnet_ipv4, name: 'one',
+        network: '192.168.42.0',
+        mask: '255.255.255.0',
+        gateway: '192.168.42.1',
+        dns_primary: '8.8.8.8',
+        dns_secondary: '8.8.4.4',
+        mtu: '1042',
+        boot_mode: :static)
+      architecture = FactoryBot.build(:architecture, name: 'x86_64')
+      medium = FactoryBot.build(:medium, name: 'CentOS mirror')
+      ptable = FactoryBot.build(:ptable, name: 'ptable')
+      operatingsystem = FactoryBot.build(:operatingsystem, name: 'Redhat',
+                                         major: 7,
+                                         architectures: [architecture],
+                                         media: [medium],
+                                         ptables: [ptable])
+      host = FactoryBot.build(:host, :managed, hostname: 'snapshot-host4static',
+                       domain: domain,
+                       subnet: subnet,
+                       architecture: architecture,
+                       medium: medium,
+                       ptable: ptable,
+                       operatingsystem: operatingsystem,
+                       interfaces: [interface])
+      define_host_params(host)
+    end
+
+    def host6
+      interface = FactoryBot.build(:nic_primary_and_provision, identifier: 'eth0',
+        mac: '00-f0-54-1a-7e-e0',
+        ip6: '2001:db8:42::42')
+      domain = FactoryBot.build(:domain, name: 'snapshotdomain.com')
+      subnet6 = FactoryBot.build(:subnet_ipv6, name: 'one', network: '2001:db8:42::')
+      architecture = FactoryBot.build(:architecture, name: 'x86_64')
+      medium = FactoryBot.build(:medium, name: 'CentOS mirror')
+      ptable = FactoryBot.build(:ptable, name: 'ptable')
+      operatingsystem = FactoryBot.build(:operatingsystem, name: 'Redhat',
+                                         major: 7,
+                                         architectures: [architecture],
+                                         media: [medium],
+                                         ptables: [ptable])
+      host = FactoryBot.build(:host, :managed, hostname: 'snapshot-host6',
+                       domain: domain,
+                       subnet6: subnet6,
+                       architecture: architecture,
+                       medium: medium,
+                       ptable: ptable,
+                       operatingsystem: operatingsystem,
+                       interfaces: [interface])
```

Comment:
This part is mostly the same between all of these. Can we extract that to a helper?

Reply:
I prefer verbose version for tests or test stubs. Helpers make code less unreadable in this case.

Reply:
Also note I have already extracted some things into `define_host_params` but other lines do not make much sense to me. It's hard to change things than and as I said less readable.

Reply:
Perhaps it could be a [trait in FactoryBot](https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#traits)? So you don't define the arch, medium, ptable and OS every time. That already saves a lot of duplication.

Reply:
That's a good idea.